libr/anal/fcn.c

```diff
@@ -61,6 +62,54 @@ R_API const char *r_anal_fcn_type_tostring(int type) {
 	return "unk";
 }
 
+#if READ_AHEAD
+// TODO: move into io :?
+static int read_ahead(RAnal *anal, ut64 addr, ut8 *buf, int len) {
+	if (len < 1) {
+		return 0;
+	}
+	static ut8 cache[1024];
+	static ut64 cache_addr = UT64_MAX;
+	const int cache_len = sizeof (cache);
+	bool isCached = (addr >= cache_addr && (addr < (cache_addr + cache_len)) && (addr + len) < (cache_addr + cache_len));
+	if (len > cache_len) {
+		return anal->iob.read_at (anal->iob.io, addr, buf, len); // double read
+	}
+	if (cache_addr == UT64_MAX || !isCached) {
+		if (len > cache_len) {
+			anal->iob.read_at (anal->iob.io, addr, buf, len);
+			memcpy (cache, buf, cache_len);
+		} else {
+			anal->iob.read_at (anal->iob.io, addr, cache, sizeof (cache));
+			memcpy (buf, cache, len);
+		}
+		cache_addr = addr;
+	} else {
+		if (isCached) {
+			if ((addr >= cache_addr) && addr < (cache_addr + cache_len) && (addr + len) < (cache_addr + cache_len)) {
+				if (addr > cache_addr + cache_len) {
+					anal->iob.read_at (anal->iob.io, addr, buf, len);
+				} else {
+					memcpy (buf, cache + (addr - cache_addr), len);
+				}
+			} else {
+				anal->iob.read_at (anal->iob.io, addr, buf, len);
+				// worth caching?
+		//		memcpy (cache, buf, cache_len);
+		//		cache_addr = addr;
+			}
+		} else {
+			// should not be reachable
```

Comment:
This isn't reachable. The conditions in force in the `else` block starting from line 87 is the inverse of `cache_addr == UT64_MAX || !isCached` i.e. `cache_addr != UT64_MAX && isCached`.

Reply:
yeah, thats why there's a comment on this :D i will clean this caching helper a bit more today